cmd/admin-obd.go

```diff
@@ -0,0 +1,399 @@
+/*
+ * MinIO Client (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/fatih/color"
+	"github.com/klauspost/compress/gzip"
+	"github.com/minio/cli"
+	cjson "github.com/minio/mc/pkg/colorjson"
+	"github.com/minio/mc/pkg/probe"
+	"github.com/minio/minio/pkg/console"
+	"github.com/minio/minio/pkg/madmin"
+)
+
+var adminOBDFlags = []cli.Flag{
+	OBDDataTypeFlag{
+		Name:   "tests",
+		Usage:  "choose OBD tests to run [" + options.String() + "]",
+		Value:  nil,
+		EnvVar: "MINIO_OBD_DATA",
+		Hidden: true,
+	},
+}
+
+var adminOBDCmd = cli.Command{
+	Name:   "obd",
+	Usage:  "run on-board diagnostics",
+	Action: mainAdminOBD,
+	Before: setGlobalsFromContext,
+	Flags:  append(adminOBDFlags, globalFlags...),
+	CustomHelpTemplate: `NAME:
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} TARGET
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+EXAMPLES:
+  1. Get server information of the 'play' MinIO server.
+     {{.Prompt}} {{.HelpName}} play/
+`,
+}
+
+type clusterOBDStruct struct {
+	Status string         `json:"status,omitempty"`
+	Error  string         `json:"error,omitempty"`
+	Info   madmin.OBDInfo `json:"obdInfo,omitempty"`
+}
+
+func (u clusterOBDStruct) String() string {
+	u.Status = ""
+	data, err := json.Marshal(u)
+	if err != nil {
+		fatalIf(probe.NewError(err), "unable to marshal into JSON.")
+	}
+	return string(data)
+}
+
+// JSON jsonifies service status message.
+func (u clusterOBDStruct) JSON() string {
+	statusJSONBytes, e := json.MarshalIndent(u, " ", "    ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+
+	return string(statusJSONBytes)
+}
+
+// checkAdminInfoSyntax - validate arguments passed by a user
+func checkAdminOBDSyntax(ctx *cli.Context) {
+	if len(ctx.Args()) == 0 || len(ctx.Args()) > 1 {
+		cli.ShowCommandHelpAndExit(ctx, "obd", 1) // last argument is exit code
+	}
+}
+
+//compress and tar obd output
+func tarGZ(c clusterOBDStruct, alias string) error {
+	filename := fmt.Sprintf("%s-obd_%s.json.gz", alias, time.Now().Format("20060102150405"))
+	f, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR, 0644)
```

Comment:
use `0666` instead to honor user umask such that the file is created with default user permissions instead of 0644